equinox/nn/_attention.py

```diff
@@ -272,59 +330,53 @@ def __call__(
             `(kv_seq_length, value_size)`.
         - `mask`: Optional mask preventing attention to certain positions. Should either
             be a JAX array of shape `(query_seq_length, kv_seq_length)`, or (for custom
-            per-head masking) `(num_heads, query_seq_length, kv_seq_length)`. A value of
-            `False` at a position indicates that position should be ignored.
+            per-head masking) `(num_heads, query_seq_length, kv_seq_length)` or
+            `causal`. A value of `False` at a position indicates that position should be
+            ignored. Use the string `causal` if you plan to use autoregressive decoding.
+        - `state`: Optional state to be passed in to the `kv_cache` callable. Used for
+            autoregressive decoding only.
         - `key`: A `jax.random.PRNGKey` used for dropout. Unused if `dropout = 0`.
             (Keyword only argument.)
         - `inference`: As [`equinox.nn.Dropout.__call__`][]. (Keyword only
             argument.)
         - `deterministic`: (Deprecated in favour of `inference`.)
-        - `process_heads`: A function that takes in the query, key, and value heads and
-            returns new query, key, and value heads. For example, this can be
-            used to implement relative positional embeddings -
-            see e.g. `RotaryPositionalEmbedding`for an example. (Keyword only argument.)
+        - `process_heads`: A function that takes in the query, key, value heads as well
+            as the current autoregressive index and returns new query, key, and value
+            heads. For example, this can be used to implement relative positional
+            embeddings - see e.g. `RotaryPositionalEmbedding`for an example.
+            (Keyword only argument.)
 
         **Returns:**
 
         A JAX array of shape `(query_seq_length, output_size)`.
         """
-
         if deterministic is not None:
             inference = deterministic
             warnings.warn(
                 "MultiheadAttention()(deterministic=...) is deprecated "
                 "in favour of MultiheadAttention()(inference=...)"
             )
 
-        query_seq_length, _ = query.shape
-        kv_seq_length, _ = key_.shape
-        kv_seq_length2, _ = value.shape
-        if kv_seq_length != kv_seq_length2:
-            # query length can be different
-            raise ValueError("key and value must both be sequences of equal length.")
+        query_seq_length, kv_seq_length = self._get_query_and_kv_seq_lengths(
+            query, key_, value
+        )
```

Comment:
Can we revert this change? I find (a) find it easier to read without all the little functions, but (b) more importantly, it's substantially easier for me to review a smaller diff.